### PR TITLE
refactor: structure verdaccio environment

### PR DESCRIPTION
**Pull request by BioPhoton**
This PR separates verdaccio and npm env logic and provides new helper to use them.

This PR introduces a new E2E setup that isolates all related part of the server, the setup code and the test files in one folder.

It includes:
- fucntion to create verdaccio environment with the following parts
  - storage folder for the packages published to verdaccio
  - `.npmrc`
- better config logit that now targets a specific directory 
- refined e2e folder structures

### Before - changes/generated files during e2e (already refactored to multiple verdaccio instances)

```sh
User/
 └── <user-name>/
     ├── .npmrc # 🔓 OS user specific npm config registry and token edited 
     └──Root/ # 👈 this is your CWD
        ├── node_modules/
        │   └── <org>
        │       └── <package-name>/... # 🔓 npm install installs into repository folder
        ├── dist/
        │   └── packages/
        │       └── <project-name>/...
        ├── tmp/
        │    └── e2e/
        │       └── registry-<port>/ # 😓 hard to debug a dynamic port
        │           ├── storage/...
        │           └── <test-name>/...
        │                └── src/...
        ├── package-lock.json # 🔓 npm install/uninstall installs into workspace root
        └── package.json # 🔓 npm install/uninstall installs into workspace root
```

### After - Changes/generated files during e2e tests

```sh
Root/ # 👈 this is your CWD
├── dist/
│   └── packages/
│       └── <project-name>/...
└── tmp/
    └── e2e/
        └── <project-name>/ # e2e setup
            ├── storage/... # npm publish/unpublish
            ├── node_modules/
            │   └── <org> 
            │       └── <package-name>/... # npm install/uninstall
            ├── __test__/...
            │   └── <file-name>/... # e2e beforeEach
            │        └── <it-block-setup>/... 
            ├── .npmrc # local npm config configured for project specific verdaccio registry 
            ├── package-lock.json # npm install/uninstall
            └── package.json # npm install/uninstall
```

### CODE REVIEW

File: tools/src/verdaccio/env.ts [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  stop: () => void;
+};
+
+export async function nxStartVerdaccioAndSetupEnv({
+  projectName,
+  port,
+  verbose = false,
+  workspaceRoot: workspaceRootDir = projectE2eScope(projectName),
+  location = 'none',
+  // reset or remove cached packages and or metadata.
```
Comment (matejchalk) [inline tools/src/verdaccio/env.ts:98]:
```suggestion
  // reset or remove cached packages and/or metadata.
```

File: tools/src/verdaccio/env.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  clear = true,
+}: StartVerdaccioAndSetupEnvOptions): Promise<VerdaccioEnvResult> {
+  // setup NPM workspace environment
```
Comment (matejchalk) [inline tools/src/verdaccio/env.ts:101]:
```suggestion
  // set up NPM workspace environment
```

File: tools/src/verdaccio/registry.ts [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+};
+
+export type VerdaccioExecuterOptions = {
+  storage?: string;
+  port?: string;
+  p?: string;
+  config?: string;
+  c?: string;
+  location: string;
+  // reset or remove cached packages and or metadata.
```
Comment (matejchalk) [inline tools/src/verdaccio/registry.ts:75]:
```suggestion
  // reset or remove cached packages and/or metadata.
```

File: tools/src/verdaccio/registry.ts [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+          }
+        },
+        onStderr: (data: string) => {
+          if (verbose) {
+            process.stdout.write(data);
+          }
+        },
+      },
+    })
+      // @TOSO reconsider this error handling
```
Comment (matejchalk) [inline tools/src/verdaccio/registry.ts:166]:
```suggestion
      // @TODO reconsider this error handling
```

File: tools/src/verdaccio/constants.ts [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import { join } from 'node:path';
+import 'verdaccio';
+
+export const START_VERDACCIO_SERVER_TARGET_NAME = 'start-verdaccio-server';
+export const START_VERDACCIO_ENV_TARGET_NAME = 'start-verdaccio-env';
+export const STOP_VERDACCIO_TARGET_NAME = 'stop-verdaccio';
+export const DEFAULT_VERDACCIO_STORAGE = join('tmp', 'verdaccio/storage');
+export const DEFAULT_VERDACCIO_CONFIG = '.verdaccio/config.yml';
```
Comment (matejchalk) [inline tools/src/verdaccio/constants.ts:8]:
```suggestion
export const DEFAULT_VERDACCIO_STORAGE = join('tmp', 'verdaccio', 'storage');
export const DEFAULT_VERDACCIO_CONFIG = join('.verdaccio', 'config.yml)';
```

File: tools/src/verdaccio/constants.ts [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import { join } from 'node:path';
+import 'verdaccio';
```
Comment (matejchalk) [inline tools/src/verdaccio/constants.ts:2]: Unused import:

```suggestion
```

File: tools/src/verdaccio/env.ts [hunk 7]
```diff
@@ -1,21 +1,24 @@
-import { execFileSync } from 'child_process';
-import { execSync } from 'node:child_process';
+// eslint-disable-next-line n/no-sync
+import { bold, gray, red } from 'ansis';
+import { execFileSync, execSync } from 'node:child_process';
```
Comment (matejchalk) [inline tools/src/verdaccio/env.ts:3]:
```suggestion
import { bold, gray, red } from 'ansis';
// eslint-disable-next-line n/no-sync
import { execFileSync, execSync } from 'node:child_process';
```

File: tools/src/verdaccio/env.ts [hunk 8]
```diff
@@ -1,21 +1,24 @@
...
 import { join } from 'node:path';
-import {
-  ensureDirectoryExists,
-  projectE2eScope,
-} from '@code-pushup/test-utils';
-import { objectToCliArgs } from '../../../packages/nx-plugin';
+import { objectToCliArgs } from '../../../packages/nx-plugin/src';
+// cant import from utils
```
Comment (matejchalk) [inline tools/src/verdaccio/env.ts:6]:
```suggestion
// can't import from utils
import { objectToCliArgs } from '../../../packages/nx-plugin/src';
```

File: tools/src/verdaccio/registry.ts [hunk 9]
```diff
@@ -1,5 +1,7 @@
-import { ConfigYaml } from '@verdaccio/types/build/configuration';
-import { executeProcess, objectToCliArgs } from '@code-pushup/utils';
+import { bold, gray, red } from 'ansis';
+import { executeProcess } from '@code-pushup/utils';
+import { objectToCliArgs } from '../../../packages/nx-plugin';
+// cant import from utils
```
Comment (matejchalk) [inline tools/src/verdaccio/registry.ts:4]:
```suggestion
// can't import from utils
import { objectToCliArgs } from '../../../packages/nx-plugin';
```

File: tools/src/verdaccio/registry.ts [hunk 10]
```diff
@@ -89,92 +91,108 @@ export function nxStartVerdaccioServer({
...
-            }`,
-          );
-          reject(error);
-        }
-        teardownTestFolder(storage);
-        throw error;
-      });
-  });
+      // in case the server dies unexpectedly clean folder
+      .catch(() => teardownTestFolder(storage))
```
Comment (matejchalk) [inline tools/src/verdaccio/registry.ts:196]: Careful not to swallow the error.

```suggestion
      .catch((error: unknown) => {
        teardownTestFolder(storage);
        throw error;
      })
```

File: e2e/nx-plugin-e2e/tests/plugin-create-nodes.e2e.test.ts [hunk 11]
```diff
@@ -34,35 +29,29 @@ describe('nx-plugin', () => {
...
-    );
+    registerPluginInWorkspace(tree, '@code-pushup/nx-plugin');
     await materializeTree(tree, cwd);
 
     const { code, projectJson } = await nxShowProjectJson(cwd, project);
     expect(code).toBe(0);
 
     expect(projectJson.targets).toStrictEqual({
       ['code-pushup--configuration']: {
-        configurations: {}, // @TODO understand why this appears. should not be here
```
[the remark this answers is not in the document]
Comment (BioPhoton) [reply e2e/nx-plugin-e2e/tests/plugin-create-nodes.e2e.test.ts:48]: Yes! It is generated by Nx plugins or other logic inside Nx, and it is normal.

**Review by matejchalk — APPROVED** [review]

**Pull request merged**